pkg/cluster/manager.go

```diff
@@ -0,0 +1,1875 @@
+	opt ScaleOutOptions,
+	skipConfirm bool,
+	optTimeout int64,
+	sshTimeout int64,
+) error {
+	metadata, err := d.meta(clusterName)
+	if err != nil {
+		return perrs.AddStack(err)
+	}
+
+	topo := metadata.GetTopology()
+	base := metadata.GetBaseMeta()
+
+	// not allowing validation errors
+	if err := topo.Validate(); err != nil {
+		return err
+	}
+
+	// Inherit existing global configuration. We must assign the inherited values before unmarshalling
+	// because some default value rely on the global options and monitored options.
+	newPart := topo.NewPart()
+
+	// The no tispark master error is ignored, as if the tispark master is removed from the topology
+	// file for some reason (manual edit, for example), it is still possible to scale-out it to make
+	// the whole topology back to normal state.
+	if err := clusterutil.ParseTopologyYaml(topoFile, newPart); err != nil &&
+		!errors.Is(perrs.Cause(err), spec.ErrNoTiSparkMaster) {
+		return err
+	}
+
+	if err := validateNewTopo(newPart); err != nil {
+		return err
+	}
+
+	// Abort scale out operation if the merged topology is invalid
+	mergedTopo := topo.MergeTopo(newPart)
+	if err := mergedTopo.Validate(); err != nil {
+		return err
+	}
+
+	if err := prepare.CheckClusterPortConflict(d.specManager, clusterName, mergedTopo); err != nil {
+		return err
+	}
+	if err := prepare.CheckClusterDirConflict(d.specManager, clusterName, mergedTopo); err != nil {
+		return err
+	}
+
+	patchedComponents := set.NewStringSet()
+	newPart.IterInstance(func(instance spec.Instance) {
+		if utils.IsExist(d.specManager.Path(clusterName, spec.PatchDirName, instance.ComponentName()+".tar.gz")) {
+			patchedComponents.Insert(instance.ComponentName())
+		}
+	})
+
+	if !skipConfirm {
+		// patchedComponents are components that have been patched and overwrited
+		if err := d.confirmTopology(clusterName, base.Version, newPart, patchedComponents); err != nil {
+			return err
+		}
+	}
+
+	sshConnProps, err := cliutil.ReadIdentityFileOrPassword(opt.IdentityFile, opt.UsePassword)
+	if err != nil {
+		return err
+	}
+
+	// Build the scale out tasks
+	t, err := buildScaleOutTask(d, clusterName, metadata, mergedTopo, opt, sshConnProps, newPart, patchedComponents, optTimeout, sshTimeout, afterDeploy, final)
+	if err != nil {
+		return err
+	}
+
+	if err := t.Execute(task.NewContext()); err != nil {
+		if errorx.Cast(err) != nil {
+			// FIXME: Map possible task errors and give suggestions.
+			return err
+		}
+		return perrs.Trace(err)
+	}
+
+	log.Infof("Scaled cluster `%s` out successfully", clusterName)
+
+	return nil
+}
+
+func (d *Manager) meta(name string) (metadata spec.Metadata, err error) {
+	exist, err := d.specManager.Exist(name)
+	if err != nil {
+		return nil, perrs.AddStack(err)
+	}
+
+	if !exist {
+		return nil, perrs.Errorf("cluster `%s` not exists", name)
+	}
+
+	metadata = d.specManager.NewMetadata()
+	err = d.specManager.Metadata(name, metadata)
+	if err != nil && !errors.Is(perrs.Cause(err), meta.ErrValidate) &&
```

Comment:
This error should be returned to the caller to decide whither ignore validation error or not, as my previous comment said...

Reply:
Reading the metadata should not failed by the Validate, the user still can call Validate explicitly if need like L1152 in cluster/manager.go.

It's a bad idea to do the check like this in #531 at many places
```
			if err != nil && !errors.Is(perrs.Cause(err), meta.ErrValidate) &&
				!errors.Is(perrs.Cause(err), spec.ErrNoTiSparkMaster) {
				return perrs.AddStack(err)
			}
```

I think it will better to let the user validate explicitly if they need instead of returning the error and you may need to ignore it.

Reply:
There is no "user" of this function.

My point is that the meta.yaml may be broken or with illegal content, due to no matter what reason, and we need to keep there's still some possibility for user to correct that with commands like `display`, `scale-in`, `edit-config` and `destroy`.

Reply:
> There is no "user" of this function.

I mean the caller of `Manager.meta` here.
So no need to check for the commands you said here.
and no need to add ignoring one more error type fore the commands when calling this function when you add one more
error type in the validate process during unmarshaling the file.

Reply:
There will always be some checks, if not for those ignoring some errors, then must be for those not ignoring errors.

Our sub commands treat the validation error differently, I'm not feeling good with the idea of ignoring some errors and recheck outside, I'm more comfortable with throwing errors to upper level and let the upper level decide what to do.

Reply:
You can treat `meta()` here as an upper level of Unmarshaling here.
the caller will not need to care about what error type to ignore.

Reply:
In `Upgrade()` the `m.meta()` is called to get the metadata of cluster, this is for example a case that validation error should not be ignored. If the topology is incorrect, upgrading will very likely make things worse.

Reply:
then you can Validate it in `Upgrade()`...

Note in master https://github.com/pingcap/tiup/blob/f1142b1c787065f3d4f2754a9b76ad26a0670f27/components/cluster/command/upgrade.go#L86, also v1.0.8

it always ignores the ErrValidate for the upgrade command(only ignore ErrValidate but not ErrNoTiSparkMaster?)

Your example just shows it's good to do it like now IMO.

Reply:
`Upgrade()` is just an example and not necessarily to be the case. So my point is always that:
 - We need to treat validation error differently in different sub commands
 - The error is raised when unmarshaling the yaml file
 - There are two ways of doing so:
    1. Ignore all validation errors when unmarshaling the file, and do validation again for those sub commands where the validation errors should not be ignored
    2. Raise any errors occurs during unmarshaling the file, and ignore validation errors for those sub commands where the validation errors shoud be ignored

I think 2 is better because ignoring errors in `m.meta()` makes the upper level caller getting less information, this raises limitation of the caller to process against errors.

If your concerns is about repeating the `errors.Is()` boolean sentence too many times, It's also okay to use a dedicated function like `IsValidationError()` to wrap them, this would be no more lines than calling `Validate()` again as it is now in `ScaleOut()`.

Reply:
`Unmarshal` should be `Unmarshal`, I think it will better to remove the `Validate` hack in `Unmarshal` but it will need more change.  It's hard to know you doing this int the `Unmarshal`.

What if you need just the truly `Unmarshal`?  (similar to the currently meta())

Reply:
Unmarshal itself has formatting validation. I'm okay to split spec validation from unmarshal, but that will lead to a problem that the validation is not atomic anymore, we'll need some wrapping to read the topology, unmarshal it and then validate it.

Actually, I don't see any situation that we need to read the topology file but not validating it's content. The "validation errors" been ignored in some sub commands are not all errors the validation process may raise, but only some topology logic error we defined in our own code.

Reply:
https://github.com/pingcap/tiup/pull/587/commits/8d8d3c8275aff976017302143e03d54e6c86204e
align with the master